components/style/animation.rs

```diff
@@ -172,100 +173,223 @@ pub enum KeyframesIterationState {
     Finite(f64, f64),
 }
 
+/// A temporary data structure used when calculating ComputedKeyframes for an
+/// animation. This data structure is used to collapse information for steps
+/// which may be spread across multiple keyframe declarations into a single
+/// instance per `start_percentage`.
+struct IntermediateComputedKeyframe {
+    declarations: PropertyDeclarationBlock,
+    timing_function: Option<TimingFunction>,
+    start_percentage: f32,
+}
+
+impl IntermediateComputedKeyframe {
+    fn new(start_percentage: f32) -> Self {
+        IntermediateComputedKeyframe {
+            declarations: PropertyDeclarationBlock::new(),
+            timing_function: None,
+            start_percentage,
+        }
+    }
+
+    /// Walk through all keyframe declarations and combine all declarations with the
+    /// same `start_percentage` into individual `IntermediateComputedKeyframe`s.
+    fn generate_for_keyframes(
+        animation: &KeyframesAnimation,
+        context: &SharedStyleContext,
+        base_style: &Arc<ComputedValues>,
+    ) -> Vec<Self> {
+        let mut intermediate_steps: Vec<Self> = Vec::with_capacity(animation.steps.len());
+        let mut current_step = IntermediateComputedKeyframe::new(0.);
+        for step in animation.steps.iter() {
+            let start_percentage = step.start_percentage.0;
+            if start_percentage != current_step.start_percentage {
+                let new_step = IntermediateComputedKeyframe::new(start_percentage);
+                intermediate_steps.push(std::mem::replace(&mut current_step, new_step));
+            }
+
+            current_step.update_from_step(step, context, base_style);
+        }
+        intermediate_steps.push(current_step);
+
+        // We should always have a first and a last step, even if these are just
+        // generated by KeyframesStepValue::ComputedValues.
+        debug_assert!(intermediate_steps.first().unwrap().start_percentage == 0.);
+        debug_assert!(intermediate_steps.last().unwrap().start_percentage == 1.);
+
+        intermediate_steps
+    }
+
+    fn update_from_step(
+        &mut self,
+        step: &KeyframesStep,
+        context: &SharedStyleContext,
+        base_style: &Arc<ComputedValues>,
```

Comment:
I think this can just be `&ComputedValues`, and same in the caller.

Reply:
Oh, nice.